Console/Command/BatchingOptimizeCommand.php

```diff
@@ -0,0 +1,366 @@
+        protected RecordBuilder         $recordBuilder,
+        protected WriterInterface       $configWriter,
+        ?string                         $name = null
+    ) {
+        parent::__construct($state, $storeNameFetcher, $name);
+    }
+
+    protected function getCommandPrefix(): string
+    {
+        return parent::getCommandPrefix() . 'batching:';
+    }
+
+    protected function getCommandName(): string
+    {
+        return 'optimize';
+    }
+
+    protected function getCommandDescription(): string
+    {
+        return "Scans some products to determine the average product record size.";
+    }
+
+    protected function getStoreArgumentDescription(): string
+    {
+        return 'ID(s) for store(s) to optimize (optional), if no store is specified, all stores will be taken into account.';
+    }
+
+    protected function getAdditionalDefinition(): array
+    {
+        return [
+            new InputOption(
+                self::LARGE_SAMPLE_OPTION,
+                '-' . self::LARGE_SAMPLE_OPTION_SHORTCUT,
+                InputOption::VALUE_NONE,
+                'Use a large sample of products (100)'
+            )
+        ];
+    }
+
+    /**
+     * @throws NoSuchEntityException|LocalizedException
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $this->input = $input;
+        $this->output = $output;
+        $this->setAreaCode();
+
+        $storeIds = $this->getStoreIds($input);
+
+        try {
+            $this->scanProductRecords($storeIds);
+        } catch (\Exception $e) {
+            $this->output->writeln('<error>' . $e->getMessage() . '</error>');
+            return CLI::RETURN_FAILURE;
+        }
+
+        return Cli::RETURN_SUCCESS;
+    }
+
+    /**
+     * @param array $storeIds
+     * @return void
+     */
+    protected function scanProductRecords(array $storeIds = []): void
+    {
+        if (count($storeIds)) {
+            foreach ($storeIds as $storeId) {
+                $this->scanProductRecordsForStore($storeId);
+            }
+        } else {
+            $this->scanProductRecordsForAllStores();
+        }
+    }
+
+    /**
+     * @return void
+     */
+    protected function scanProductRecordsForAllStores(): void
+    {
+        $storeIds = array_keys($this->storeManager->getStores());
+
+        foreach ($storeIds as $storeId) {
+            $this->scanProductRecordsForStore($storeId);
+        }
+    }
+
+    /**
+     * @param int $storeId
+     * @return void
+     */
+    protected function scanProductRecordsForStore(int $storeId): void
+    {
+        $storeName = $this->storeNameFetcher->getStoreName($storeId);
+
+        if (!$this->configHelper->isIndexingEnabled($storeId)) {
+            $this->output->writeln('<info>Indexing is disabled for store ' . $storeName . '</info>');
+            return;
+        }
+
+        if (!isset($this->storeCounts[$storeId])) {
+            $this->setStoreCounts($storeId);
```

Comment:
This method can throw exceptions but we don't catch or call out - maybe add some `@throws` statements in the calling method?